# Implementation of baseline using Artificial Neural Network with Original + Win-loss + feature vectors feature set and EPL

Necessary files: logistic_regression_functions.py, ann_functions.py, testing_functions.py, win_loss_functions.py, feature_vectors_functions, EPL_17.csv, EPL_18.csv, EPL_19.csv, A_EPL.csv, H_EPL.csv, A_EPL_before_threshold.csv, H_EPL_before_threshold.csv

In [ ]:
import logistic_regression_functions
import ann_functions
import testing_functions
import win_loss_functions
import feature_vectors_functions
import pandas as pd
from datetime import datetime
import requests
import numpy as np
import tensorflow as tf
from tensorflow.keras import Model
from tensorflow.keras.layers import *
from tqdm import tqdm

The baseline is implemented according to the paper An Improved Prediction System for Football a Match Result bu C. P. Igiri and E. O. Nwachukwu from 2014. The original features as described in the paper are used and the model is evaluated on the EPL dataset.

The authors used only one season for training, so I did the same. I used seasons 2016/17 for training season 2017/18 as validation set and the second half of season 2018/19 for testing. The validation set is used to select the best checkpoint of the training according to the validation accuracy.

In [ ]:
X_train_win_loss, y_train = win_loss_functions.create_data(['EPL_17.csv'], skip_rounds = 6, return_names=True)
results_val_win_loss, matches_per_round = win_loss_functions.create_data_single('EPL_18.csv', ['EPL_17.csv'],
                                                                       return_names=True, skip_rounds = 6)
# Dates are returned as well for dividing testing season into slices
results_test_win_loss, matches_per_round = win_loss_functions.create_data_single('EPL_19.csv', ['EPL_17.csv', 'EPL_18.csv'],
                                return_dates=True, skip_rounds = 6, return_names=True)

Processing EPL_17.csv season file.


In [ ]:
results_train_originals, matches_per_round = ann_functions.create_data_single('EPL_17.csv', 13, 17,
                                   logistic_regression_functions.team_names_map_epl,
                                   logistic_regression_functions.secondary_team_names_map_epl)
results_val_originals, matches_per_round = ann_functions.create_data_single('EPL_18.csv', 13, 18,
                                  logistic_regression_functions.team_names_map_epl,
                                  logistic_regression_functions.secondary_team_names_map_epl)
# Dates are returned as well for dividing testing season into slices
results_test_originals, matches_per_round = ann_functions.create_data_single('EPL_19.csv', 13, 19,
                                  logistic_regression_functions.team_names_map_epl,
                                  logistic_regression_functions.secondary_team_names_map_epl,
                                  return_dates=True)
X_train_originals = results_train_originals.drop('FTR', axis=1)
y_train_originals = results_train_originals['FTR']

In [ ]:
# Concatenating the two feature sets together
X_train_mix = pd.concat([X_train_originals, X_train_win_loss], axis=1)
results_val_mix = pd.concat([results_val_originals, results_val_win_loss], axis=1)
results_test_mix = pd.concat([results_test_originals, results_test_win_loss], axis=1)
X_train_mix = X_train_mix.loc[:,~X_train_mix.columns.duplicated()].copy()
results_val_mix = results_val_mix.loc[:,~results_val_mix.columns.duplicated()].copy()
results_test_mix = results_test_mix.loc[:,~results_test_mix.columns.duplicated()].copy()

In [ ]:
A = pd.read_csv('A_EPL.csv')
H = pd.read_csv('H_EPL.csv')
A_before_threshold = pd.read_csv('A_EPL_before_threshold.csv')
H_before_threshold = pd.read_csv('H_EPL_before_threshold.csv')

In [ ]:
# Adding the feature vectors to the features
X_train = feature_vectors_functions.add_feature_vector(X_train_mix, A, H)
results_val = feature_vectors_functions.add_feature_vector(results_val_mix, A, H)
results_test = feature_vectors_functions.add_feature_vector(results_test_mix, A, H)

In [ ]:
X_val = results_val.drop('FTR', axis=1)
y_val = results_val['FTR']
# The classes need to go from 0 to 2 not from -1 to 1.
y_train += 1
y_val += 1

They didnt say what ANN architecture they used. I used 3 dense layers with 64, 32 and 16 neurons. I used dropout to overcome overfitting.

I always let the model train for 100 epochs and took the weights from the best epoch according to validation accuracy.

To use all the data available, the first half of the testing season was added to the training data.

In [ ]:
# Some rounds in the beginning are ignored, this is the correct index
# of the start of the second half of the season
start_test_index = 13 * matches_per_round

In [ ]:
X_test_to_append, y_test_to_append = testing_functions.prepare_test_to_append(results_test,
                                                                              start_test_index)
y_test_to_append += 1

In [ ]:
# Adding 1st half of testing season to the training data
X_train = pd.concat([X_train, X_test_to_append])
y_train = pd.concat([y_train, y_test_to_append])
# Rounds of the testing dataset
slices = testing_functions.get_slices(results_test, matches_per_round,
                                      start_test_index)
weighted_sum = 0
sum = 0
for slc in tqdm(slices):
  # Creating the model
  model = ann_functions.func_model((X_train.shape[1],))
  model.compile(optimizer='adam',
                 loss='sparse_categorical_crossentropy',
                 metrics=['accuracy'])
  mc = tf.keras.callbacks.ModelCheckpoint('./weights_model.h5',
                                     monitor='val_accuracy',
                                     save_weights_only=True,
                                     save_best_only=True)
  X_test = slc.drop(['FTR', 'Date'], axis=1)
  y_test = slc['FTR']
  y_test += 1
  # Train the model
  history = model.fit(X_train, y_train,
                    epochs=100,
                    batch_size=8,
                    validation_data=(X_val, y_val),
                    verbose = 0,
                    callbacks=[mc])
  # Load the best checkpoint
  model.load_weights('weights_model.h5')
  weighted_sum += (model.evaluate(X_test, y_test)[1] * len(y_test))
  sum += len(y_test)
  # Add the round to the training dataset
  X_train = pd.concat([X_train, X_test])
  y_train = pd.concat([y_train, y_test])
print('')
print(weighted_sum / sum)

  0%|          | 0/15 [00:00<?, ?it/s]

1/1 [==============================] - 0s 183ms/step - loss: 1.0225 - accuracy: 0.6154


  7%|▋         | 1/15 [00:18<04:19, 18.56s/it]

1/1 [==============================] - 0s 163ms/step - loss: 1.6011 - accuracy: 0.4286


 13%|█▎        | 2/15 [00:39<04:23, 20.25s/it]

1/1 [==============================] - 0s 158ms/step - loss: 0.7562 - accuracy: 0.7273


 20%|██        | 3/15 [00:58<03:53, 19.44s/it]

1/1 [==============================] - 0s 140ms/step - loss: 1.3404 - accuracy: 0.5000


 27%|██▋       | 4/15 [01:16<03:28, 18.95s/it]

1/1 [==============================] - 0s 157ms/step - loss: 1.0868 - accuracy: 0.5455


 33%|███▎      | 5/15 [01:38<03:18, 19.82s/it]

1/1 [==============================] - 0s 156ms/step - loss: 0.8048 - accuracy: 0.6667


 40%|████      | 6/15 [01:59<03:03, 20.37s/it]

1/1 [==============================] - 0s 136ms/step - loss: 0.9571 - accuracy: 0.6250


 47%|████▋     | 7/15 [02:20<02:45, 20.69s/it]

1/1 [==============================] - 0s 151ms/step - loss: 0.8178 - accuracy: 0.6471


 53%|█████▎    | 8/15 [03:02<03:12, 27.52s/it]

1/1 [==============================] - 0s 153ms/step - loss: 0.8813 - accuracy: 0.5333


 60%|██████    | 9/15 [03:24<02:33, 25.63s/it]

1/1 [==============================] - 0s 171ms/step - loss: 0.5765 - accuracy: 0.7500


 67%|██████▋   | 10/15 [03:45<02:01, 24.22s/it]

1/1 [==============================] - 0s 142ms/step - loss: 0.6448 - accuracy: 0.8182


 73%|███████▎  | 11/15 [04:06<01:33, 23.34s/it]

1/1 [==============================] - 0s 141ms/step - loss: 1.3722 - accuracy: 0.4545


 80%|████████  | 12/15 [04:48<01:26, 28.97s/it]

1/1 [==============================] - 0s 142ms/step - loss: 1.9027 - accuracy: 0.4167


 87%|████████▋ | 13/15 [05:09<00:53, 26.65s/it]

1/1 [==============================] - 0s 164ms/step - loss: 1.0321 - accuracy: 0.5833


 93%|█████████▎| 14/15 [05:31<00:25, 25.02s/it]

1/1 [==============================] - 0s 159ms/step - loss: 1.9928 - accuracy: 0.4545


100%|██████████| 15/15 [05:53<00:00, 23.54s/it]


0.5842105350996318


The testing accuracy is 58.42%.

Experimenting with feature vectors extracted before applying the threshold.

In [ ]:
# y_train was changed, load it again
X_train_win_loss, y_train = win_loss_functions.create_data(['EPL_17.csv'], skip_rounds = 6, return_names=True)

Processing EPL_17.csv season file.


In [ ]:
# Adding the feature vectors to the features
X_train = feature_vectors_functions.add_feature_vector(X_train_mix, A_before_threshold, H_before_threshold)
results_val = feature_vectors_functions.add_feature_vector(results_val_mix, A_before_threshold, H_before_threshold)
results_test = feature_vectors_functions.add_feature_vector(results_test_mix, A_before_threshold, H_before_threshold)

In [ ]:
X_val = results_val.drop('FTR', axis=1)
y_val = results_val['FTR']
# The classes need to go from 0 to 2 not from -1 to 1.
y_train += 1
y_val += 1

In [ ]:
X_test_to_append, y_test_to_append = testing_functions.prepare_test_to_append(results_test,
                                                                              start_test_index)
y_test_to_append += 1

In [ ]:
# Adding 1st half of testing season to the training data
X_train = pd.concat([X_train, X_test_to_append])
y_train = pd.concat([y_train, y_test_to_append])
# Rounds of the testing dataset
slices = testing_functions.get_slices(results_test, matches_per_round,
                                      start_test_index)
weighted_sum = 0
sum = 0
for slc in tqdm(slices):
  # Creating the model
  model = ann_functions.func_model((X_train.shape[1],))
  model.compile(optimizer='adam',
                 loss='sparse_categorical_crossentropy',
                 metrics=['accuracy'])
  mc = tf.keras.callbacks.ModelCheckpoint('./weights_model.h5',
                                     monitor='val_accuracy',
                                     save_weights_only=True,
                                     save_best_only=True)
  X_test = slc.drop(['FTR', 'Date'], axis=1)
  y_test = slc['FTR']
  y_test += 1
  # Train the model
  history = model.fit(X_train, y_train,
                    epochs=100,
                    batch_size=8,
                    validation_data=(X_val, y_val),
                    verbose = 0,
                    callbacks=[mc])
  # Load the best checkpoint
  model.load_weights('weights_model.h5')
  weighted_sum += (model.evaluate(X_test, y_test)[1] * len(y_test))
  sum += len(y_test)
  # Add the round to the training dataset
  X_train = pd.concat([X_train, X_test])
  y_train = pd.concat([y_train, y_test])
print('')
print(weighted_sum / sum)

  0%|          | 0/15 [00:00<?, ?it/s]

1/1 [==============================] - 0s 157ms/step - loss: 1.0266 - accuracy: 0.6923


  7%|▋         | 1/15 [00:21<05:00, 21.49s/it]

1/1 [==============================] - 0s 160ms/step - loss: 1.1723 - accuracy: 0.3571


 13%|█▎        | 2/15 [00:42<04:39, 21.50s/it]

1/1 [==============================] - 0s 180ms/step - loss: 0.8321 - accuracy: 0.8182


 20%|██        | 3/15 [01:04<04:18, 21.52s/it]

1/1 [==============================] - 0s 157ms/step - loss: 1.2775 - accuracy: 0.5833


 27%|██▋       | 4/15 [01:26<03:56, 21.50s/it]

1/1 [==============================] - 0s 175ms/step - loss: 0.9756 - accuracy: 0.5455


 33%|███▎      | 5/15 [01:46<03:32, 21.28s/it]

1/1 [==============================] - 0s 171ms/step - loss: 0.8864 - accuracy: 0.6667


 40%|████      | 6/15 [02:08<03:12, 21.36s/it]

1/1 [==============================] - 0s 157ms/step - loss: 0.8389 - accuracy: 0.6250


 47%|████▋     | 7/15 [02:30<02:52, 21.54s/it]

1/1 [==============================] - 0s 170ms/step - loss: 0.8160 - accuracy: 0.6471


 53%|█████▎    | 8/15 [03:12<03:16, 28.04s/it]

1/1 [==============================] - 0s 162ms/step - loss: 0.7141 - accuracy: 0.6667


 60%|██████    | 9/15 [03:35<02:38, 26.47s/it]

1/1 [==============================] - 0s 154ms/step - loss: 0.5466 - accuracy: 0.8333


 67%|██████▋   | 10/15 [04:17<02:36, 31.25s/it]

1/1 [==============================] - 0s 173ms/step - loss: 0.8329 - accuracy: 0.6364


 73%|███████▎  | 11/15 [04:40<01:55, 28.93s/it]

1/1 [==============================] - 0s 163ms/step - loss: 1.0822 - accuracy: 0.3636


 80%|████████  | 12/15 [05:23<01:38, 32.93s/it]

1/1 [==============================] - 0s 171ms/step - loss: 1.1527 - accuracy: 0.4167


 87%|████████▋ | 13/15 [06:05<01:11, 35.68s/it]

1/1 [==============================] - 0s 154ms/step - loss: 0.9592 - accuracy: 0.6667


 93%|█████████▎| 14/15 [06:47<00:37, 37.59s/it]

1/1 [==============================] - 0s 158ms/step - loss: 1.0635 - accuracy: 0.3636


100%|██████████| 15/15 [07:29<00:00, 29.93s/it]


0.5947368480657277


The testing accuracy is 59.47%.

Experimenting with feature vectors including home feature vectors for the away teams and away feature vectors for the home teams.

In [ ]:
# y_train was changed, load it again
X_train_win_loss, y_train = win_loss_functions.create_data(['EPL_17.csv'], skip_rounds = 6, return_names=True)

Processing EPL_17.csv season file.


In [ ]:
# Adding the feature vectors to the features
X_train = feature_vectors_functions.add_feature_vector(X_train_mix, A, H, include_all=True)
results_val = feature_vectors_functions.add_feature_vector(results_val_mix, A, H, include_all=True)
results_test = feature_vectors_functions.add_feature_vector(results_test_mix, A, H, include_all=True)

In [ ]:
X_val = results_val.drop('FTR', axis=1)
y_val = results_val['FTR']
# The classes need to go from 0 to 2 not from -1 to 1.
y_train += 1
y_val += 1

In [ ]:
X_test_to_append, y_test_to_append = testing_functions.prepare_test_to_append(results_test,
                                                                              start_test_index)
y_test_to_append += 1

In [ ]:
# Adding 1st half of testing season to the training data
X_train = pd.concat([X_train, X_test_to_append])
y_train = pd.concat([y_train, y_test_to_append])
# Rounds of the testing dataset
slices = testing_functions.get_slices(results_test, matches_per_round,
                                      start_test_index)
weighted_sum = 0
sum = 0
for slc in tqdm(slices):
  # Creating the model
  model = ann_functions.func_model((X_train.shape[1],))
  model.compile(optimizer='adam',
                 loss='sparse_categorical_crossentropy',
                 metrics=['accuracy'])
  mc = tf.keras.callbacks.ModelCheckpoint('./weights_model.h5',
                                     monitor='val_accuracy',
                                     save_weights_only=True,
                                     save_best_only=True)
  X_test = slc.drop(['FTR', 'Date'], axis=1)
  y_test = slc['FTR']
  y_test += 1
  # Train the model
  history = model.fit(X_train, y_train,
                    epochs=100,
                    batch_size=8,
                    validation_data=(X_val, y_val),
                    verbose = 0,
                    callbacks=[mc])
  # Load the best checkpoint
  model.load_weights('weights_model.h5')
  weighted_sum += (model.evaluate(X_test, y_test)[1] * len(y_test))
  sum += len(y_test)
  # Add the round to the training dataset
  X_train = pd.concat([X_train, X_test])
  y_train = pd.concat([y_train, y_test])
print('')
print(weighted_sum / sum)

  0%|          | 0/15 [00:00<?, ?it/s]

1/1 [==============================] - 0s 117ms/step - loss: 1.5367 - accuracy: 0.6923


  7%|▋         | 1/15 [00:21<05:05, 21.79s/it]

1/1 [==============================] - 0s 131ms/step - loss: 1.5242 - accuracy: 0.4286


 13%|█▎        | 2/15 [00:38<04:05, 18.86s/it]

1/1 [==============================] - 0s 114ms/step - loss: 0.8077 - accuracy: 0.6364


 20%|██        | 3/15 [00:59<03:59, 19.95s/it]

1/1 [==============================] - 0s 118ms/step - loss: 1.2207 - accuracy: 0.5833


 27%|██▋       | 4/15 [01:21<03:44, 20.45s/it]

1/1 [==============================] - 0s 125ms/step - loss: 0.9890 - accuracy: 0.6364


 33%|███▎      | 5/15 [01:42<03:27, 20.74s/it]

1/1 [==============================] - 0s 134ms/step - loss: 1.1251 - accuracy: 0.5833


 40%|████      | 6/15 [02:00<02:58, 19.83s/it]

1/1 [==============================] - 0s 115ms/step - loss: 0.7776 - accuracy: 0.6875


 47%|████▋     | 7/15 [02:18<02:34, 19.30s/it]

1/1 [==============================] - 0s 118ms/step - loss: 0.9146 - accuracy: 0.6471


 53%|█████▎    | 8/15 [02:37<02:14, 19.26s/it]

1/1 [==============================] - 0s 120ms/step - loss: 0.8178 - accuracy: 0.5333


 60%|██████    | 9/15 [02:59<01:59, 19.96s/it]

1/1 [==============================] - 0s 125ms/step - loss: 0.6695 - accuracy: 0.6667


 67%|██████▋   | 10/15 [03:20<01:41, 20.35s/it]

1/1 [==============================] - 0s 129ms/step - loss: 0.7223 - accuracy: 0.6364


 73%|███████▎  | 11/15 [03:41<01:22, 20.62s/it]

1/1 [==============================] - 0s 120ms/step - loss: 1.9676 - accuracy: 0.5455


 80%|████████  | 12/15 [04:02<01:02, 20.80s/it]

1/1 [==============================] - 0s 137ms/step - loss: 1.3803 - accuracy: 0.4167


 87%|████████▋ | 13/15 [04:23<00:41, 20.62s/it]

1/1 [==============================] - 0s 114ms/step - loss: 0.9793 - accuracy: 0.5833


 93%|█████████▎| 14/15 [04:43<00:20, 20.53s/it]

1/1 [==============================] - 0s 125ms/step - loss: 1.2018 - accuracy: 0.4545


100%|██████████| 15/15 [05:04<00:00, 20.31s/it]


0.5842105286686043


The testing accuracy is 58.42%.

Experimenting with feature vectors extracted before applying the threshold and including home feature vectors for the away teams and away feature vectors for the home teams.

In [ ]:
# y_train was changed, load it again
X_train_win_loss, y_train = win_loss_functions.create_data(['EPL_17.csv'], skip_rounds = 6, return_names=True)

Processing EPL_17.csv season file.


In [ ]:
# Adding the feature vectors to the features
X_train = feature_vectors_functions.add_feature_vector(X_train_mix, A_before_threshold, H, include_all=True)
results_val = feature_vectors_functions.add_feature_vector(results_val_mix, A_before_threshold, H_before_threshold, include_all=True)
results_test = feature_vectors_functions.add_feature_vector(results_test_mix, A_before_threshold, H_before_threshold, include_all=True)

In [ ]:
X_val = results_val.drop('FTR', axis=1)
y_val = results_val['FTR']
# The classes need to go from 0 to 2 not from -1 to 1.
y_train += 1
y_val += 1

In [ ]:
X_test_to_append, y_test_to_append = testing_functions.prepare_test_to_append(results_test,
                                                                              start_test_index)
y_test_to_append += 1

In [ ]:
# Adding 1st half of testing season to the training data
X_train = pd.concat([X_train, X_test_to_append])
y_train = pd.concat([y_train, y_test_to_append])
# Rounds of the testing dataset
slices = testing_functions.get_slices(results_test, matches_per_round,
                                      start_test_index)
weighted_sum = 0
sum = 0
for slc in tqdm(slices):
  # Creating the model
  model = ann_functions.func_model((X_train.shape[1],))
  model.compile(optimizer='adam',
                 loss='sparse_categorical_crossentropy',
                 metrics=['accuracy'])
  mc = tf.keras.callbacks.ModelCheckpoint('./weights_model.h5',
                                     monitor='val_accuracy',
                                     save_weights_only=True,
                                     save_best_only=True)
  X_test = slc.drop(['FTR', 'Date'], axis=1)
  y_test = slc['FTR']
  y_test += 1
  # Train the model
  history = model.fit(X_train, y_train,
                    epochs=100,
                    batch_size=8,
                    validation_data=(X_val, y_val),
                    verbose = 0,
                    callbacks=[mc])
  # Load the best checkpoint
  model.load_weights('weights_model.h5')
  weighted_sum += (model.evaluate(X_test, y_test)[1] * len(y_test))
  sum += len(y_test)
  # Add the round to the training dataset
  X_train = pd.concat([X_train, X_test])
  y_train = pd.concat([y_train, y_test])
print('')
print(weighted_sum / sum)

  0%|          | 0/15 [00:00<?, ?it/s]

1/1 [==============================] - 0s 150ms/step - loss: 1.4662 - accuracy: 0.6923


  7%|▋         | 1/15 [00:17<04:09, 17.81s/it]

1/1 [==============================] - 0s 144ms/step - loss: 1.2101 - accuracy: 0.3571


 13%|█▎        | 2/15 [00:39<04:18, 19.88s/it]

1/1 [==============================] - 0s 143ms/step - loss: 0.9029 - accuracy: 0.6364


 20%|██        | 3/15 [01:00<04:06, 20.54s/it]

1/1 [==============================] - 0s 143ms/step - loss: 1.6219 - accuracy: 0.5833


 27%|██▋       | 4/15 [01:18<03:34, 19.53s/it]

1/1 [==============================] - 0s 136ms/step - loss: 1.0340 - accuracy: 0.5455


 33%|███▎      | 5/15 [01:40<03:23, 20.31s/it]

1/1 [==============================] - 0s 131ms/step - loss: 0.8570 - accuracy: 0.6667


 40%|████      | 6/15 [02:01<03:05, 20.66s/it]

1/1 [==============================] - 0s 132ms/step - loss: 0.9026 - accuracy: 0.5625


 47%|████▋     | 7/15 [02:20<02:40, 20.08s/it]

1/1 [==============================] - 0s 134ms/step - loss: 0.9967 - accuracy: 0.5882


 53%|█████▎    | 8/15 [02:41<02:23, 20.48s/it]

1/1 [==============================] - 0s 144ms/step - loss: 0.5409 - accuracy: 0.7333


 60%|██████    | 9/15 [03:03<02:04, 20.74s/it]

1/1 [==============================] - 0s 140ms/step - loss: 0.5088 - accuracy: 0.8333


 67%|██████▋   | 10/15 [03:24<01:44, 20.92s/it]

1/1 [==============================] - 0s 131ms/step - loss: 0.8094 - accuracy: 0.8182


 73%|███████▎  | 11/15 [03:44<01:22, 20.68s/it]

1/1 [==============================] - 0s 133ms/step - loss: 1.2660 - accuracy: 0.4545


 80%|████████  | 12/15 [04:05<01:01, 20.65s/it]

1/1 [==============================] - 0s 144ms/step - loss: 1.6308 - accuracy: 0.3333


 87%|████████▋ | 13/15 [04:26<00:41, 20.86s/it]

1/1 [==============================] - 0s 137ms/step - loss: 0.9361 - accuracy: 0.5833


 93%|█████████▎| 14/15 [04:48<00:21, 21.12s/it]

1/1 [==============================] - 0s 134ms/step - loss: 1.0748 - accuracy: 0.3636


100%|██████████| 15/15 [05:08<00:00, 20.60s/it]


0.5842105311782737


The testing accuracy is 58.42%.